### ISSUE TITLE
The error of `torch.nn.SyncBatchNorm.convert_sync_batchnorm`

### ISSUE THREAD

**Issue by as754770178** [pytorch/pytorch#39643]
```
def model_parallel(model, optimizer, gpu_ids, dist_config=None):
    torch.cuda.set_device(gpu_ids[0])
    model = model.cuda(gpu_ids[0])

    if isinstance(dist_config, DistConfig) \
            and dist_config.multiprocessing_distributed and dist_config.sync_bn:
      model = torch.nn.SyncBatchNorm.convert_sync_batchnorm(model)

  optimizer, remain_optimizer = get_optimizer(model, optimizer)
  if isinstance(dist_config, DistConfig) and dist_config.dist_url is not None:
    if dist_config.multiprocessing_distributed:

      dist_config.world_size = dist_config.world_size * dist_config.ngpus_per_node
      dist_config.rank = int(dist_config.rank) * int(dist_config.ngpus_per_node) + gpu_ids[0]

    _dist.init_process_group(backend=dist_config.dist_backend,
                             init_method=dist_config.dist_url,
                             world_size=dist_config.world_size,
                             rank=dist_config.rank)
    set_dh_param(BulitInParams.NODE_RANK, dist_config.rank)

    model = torch.nn.parallel.DistributedDataParallel(model, device_ids=gpu_ids,
                                                      find_unused_parameters=True)
  elif gpu_ids is not None and len(gpu_ids) > 1:
    model = torch.nn.DataParallel(model, device_ids=gpu_ids)

  return model, optimizer, remain_optimizer
```
multiprocessing_distributed mean that I use `torch.multiprocessing`, the code as 
```
mp.spawn(
        self._train,
        nprocs=len(gpu_ids))
```

When I exec `model = torch.nn.SyncBatchNorm.convert_sync_batchnorm(model)`, the weight of bn still is 1, bias of bn still is 0.

**Comment by as754770178** [comment]
I view the parameter of bn, the training is True, require_grad is True, but grad_fn is None.

**Comment by SsnL** [comment]
> When I exec model = torch.nn.SyncBatchNorm.convert_sync_batchnorm(model), the weight of bn still is 1, bias of bn still is 0.

What do you expect otherwise? The function just replaces regular batch norm layers to SyncBatchNorm layers. It should neither change the parameters nor mess with computation graph (e.g., grad_fn).

**Comment by as754770178** [comment]
I find in the forward of model, the `_grad` of bn's weight is not None. But in optimizer's param_groups,  the `_grad` of bn's weight is None.

**Comment by SsnL** [comment]
@as754770178 Could you post a script that reproduces the unexpected behavior? I'm not sure I understand the problem.

**Comment by as754770178** [comment]
I guess the error due to the order of create optimizer and exec SyncBatchNorm, so I change 
```
if isinstance(dist_config, DistConfig) \
            and dist_config.multiprocessing_distributed and dist_config.sync_bn:
      model = torch.nn.SyncBatchNorm.convert_sync_batchnorm(model)

  optimizer, remain_optimizer = get_optimizer(model, optimizer)
```
to 
```
optimizer, remain_optimizer = get_optimizer(model, optimizer)
if isinstance(dist_config, DistConfig) \
            and dist_config.multiprocessing_distributed and dist_config.sync_bn:
      model = torch.nn.SyncBatchNorm.convert_sync_batchnorm(model)
```

But in optimizer's param_groups, the _grad of bn's weight still is None.

**Comment by SsnL** [comment]
> But in optimizer's param_groups, the _grad of bn's weight still is None.

Do you mean that the optimizer doesn't optimize the BN's weights?

The first is correct. You should always build optimizer **after** building the model.

**Comment by as754770178** [comment]
The code is big. I mean in `/usr/local/lib/python3.6/site-packages/torch/optim/sgd.py`
```
    def step(self, closure=None):
        """Performs a single optimization step.

        Arguments:
            closure (callable, optional): A closure that reevaluates the model
                and returns the loss.
        """
        loss = None
        if closure is not None:
            loss = closure()

        for group in self.param_groups:
            weight_decay = group['weight_decay']
            momentum = group['momentum']
            dampening = group['dampening']
            nesterov = group['nesterov']

            for p in group['params']:
                if p.grad is None:
                    continue
```
I view the grad of bn's weight is None.

But in the forward of model,
```
    def forward(self, *inputs):
        x, *other = inputs
        b1, b2, b3 = self.bone(x)
        out1, out2, out3 = self.head(b1, b2, b3)
        return [out1, out2, out3]
```
I check the same bn layer, the grad of weight in bn layer is not None.
I want to know what will cause this？

**Comment by as754770178** [comment]
I only delete 
```
if isinstance(dist_config, DistConfig) \
            and dist_config.multiprocessing_distributed and dist_config.sync_bn:
      model = torch.nn.SyncBatchNorm.convert_sync_batchnorm(model)
```

Then I check the grad of bn's layer is not None.

**Comment by as754770178** [comment]
and I check the parameter in opimizer and in model is the same object.
In optimizer, `/usr/local/lib/python3.6/site-packages/torch/optim/sgd.py`,
```
for group in self.param_groups:
            weight_decay = group['weight_decay']
            momentum = group['momentum']
            dampening = group['dampening']
            nesterov = group['nesterov']
            print(id(group['params'][1])) # get the id of parameter in optimizer
            for p in group['params']:
                if p.grad is None:
                    continue
```
In model, 
```
def forward(self, *inputs):
        x, *other = inputs
        b1, b2, b3 = self.bone(x)
        out1, out2, out3 = self.head(b1, b2, b3)
        id(list(self.parameters())[1])  # get the id of parameter in model
        return [out1, out2, out3]
```

If I exec `model = torch.nn.SyncBatchNorm.convert_sync_batchnorm(model)`, the id of paramter is different between optimizer and model. otherwise, the id is same between optimizer and model

**Comment by as754770178** [comment]
sorry, I fix the issue that I create optimizer before exec `model = torch.nn.SyncBatchNorm.convert_sync_batchnorm(model)`
